server/src/main/java/org/opensearch/cluster/routing/allocation/decider/NodeOverloadAwareAllocationDecider.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing.allocation.decider;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.routing.RoutingNode;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.settings.Setting.Property;
+
+import java.util.function.BiPredicate;
+
+/**
+ * This {@link NodeOverloadAwareAllocationDecider} controls shard over-allocation
+ * due to node failures or otherwise on the surviving nodes
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.provisioned_capacity: N
+ * </pre>
+ * <p>
+ * and prevent allocation on the surviving nodes of the under capacity cluster
+ * based on oveload factor defined as a percentage by
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.factor: X
+ * </pre>
+ */
+public class NodeOverloadAwareAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "overload_awareness";
+
+    public static final Setting<Integer> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING =
+        Setting.intSetting("cluster.routing.allocation.overload_awareness.provisioned_capacity", -1, -1,
+            Property.Dynamic, Property.NodeScope);
+    public static final Setting<Integer> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_FACTOR_SETTING =
+        Setting.intSetting("cluster.routing.allocation.overload_awareness.factor", 50, -1, Property.Dynamic,
+            Property.NodeScope);
+    public static final Setting<Boolean> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_ALLOW_UNASSIGNED_PRIMARIES_SETTING =
+        Setting.boolSetting("cluster.routing.allocation.overload_awareness.allow_unassigned_primaries",
+            true, Setting.Property.Dynamic, Property.NodeScope);
```

Comment:
Why do we really need this explicit setting? Shouldn’t decider be able to detect based on cluster state, such as node local copy? For new index creation, the shard distribution would be even anyways?

Reply:
if the limit is already breached, we still want new index creation to go through assigning the primaries as an exception to the overload limit

Reply:
> For new index creation, the shard distribution would be even anyways?

I don't think that is guaranteed, on the contrary node with least weight is more likely to assign most or all primary shards

Reply:
> we still want new index creation to go through assigning the primaries as an exception to the overload limit

I agree, and that could be the default behaviour i suppose.

>on the contrary node with least weight is more likely to assign most or all primary shards

This implies that allocation decision is anyways preferring the nodes with relatively less number of shards. And this setting is actually not avoiding any distribution overload for newly created shards.

Wondering if this setting can be unintentionally more intrusive to customer workload, than the wins it actually offers.

Reply:
Have explained it https://github.com/opensearch-project/OpenSearch/pull/1149#discussion_r705311241